File: 섹션08. A형 문제 풀이 (~2022 상반기)/29. 술래잡기 체스 (코드트리, 2020 상반기 오전 2번).cpp
#include <stdio.h>

#define MAX (4 + 2)
#define MAX_H (16 + 5) // 도둑말의 수

int T;

int MAP[MAX][MAX];

struct CHESS
{
	int r;
	int c;
	int dir;
	bool dead;
};

CHESS chess[MAX_H];
CHESS tagger;

int maxAnswer;

// -, ↑, ↖, ←, ↙, ↓, ↘, →, ↗
int dr[] = { 0, -1, -1, 0, 1, 1, 1, 0, -1 };
int dc[] = { 0, 0, -1, -1, -1, 0, 1, 1, 1 };

void input()
{
	for (int r = 0; r < 4; r++)
	{
		for (int c = 0; c < 4; c++)
		{
			int index, dir;

			scanf("%d %d", &index, &dir);

			MAP[r][c] = index;

			chess[index].r = r;
			chess[index].c = c;
			chess[index].dir = dir;
			chess[index].dead = false;
		}
	}
}

void printStatus(CHESS tagger, int map[MAX][MAX], CHESS chess[MAX_H], int score) // for debug
{
	printf("score : %d\n", score);
	printf("tagger : r %d, c %d, dir %d\n", tagger.r, tagger.c, tagger.dir);

	printf("live chess : ");
	for (int i = 1; i <= 16; i++)
		if (chess[i].dead == false) printf("%d, ", i);
	putchar('\n');

	for (int r = 0; r < 4; r++)
	{
		for (int c = 0; c < 4; c++)
		{
			if (r == tagger.r && c == tagger.c) printf("(%d, %d) ", -1, tagger.dir);
			else printf("(%d, %d) ", map[r][c], chess[map[r][c]].dir);
		}
		putchar('\n');
	}
	putchar('\n');
}

void DFS(CHESS prevTagger, int prevMap[MAX][MAX], CHESS prevChess[MAX_H], int score)
{
	int tmpMAP[MAX][MAX] = { 0 };
	CHESS tmpChess[MAX_H] = { 0 };
	CHESS pt = prevTagger;

	// copy
	for (int r = 0; r < 4; r++)
		for (int c = 0; c < 4; c++)
			tmpMAP[r][c] = prevMap[r][c];

	for (int i = 1; i <= 16; i++) tmpChess[i] = prevChess[i];

	int deadIndex = tmpMAP[pt.r][pt.c];

	// 이미 잡힌 도둑말인 경우
	if (tmpChess[deadIndex].dead == true) return;

	// 술래 방향 변경
	pt.dir = tmpChess[deadIndex].dir;

	// 도둑말 사망 처리
	tmpChess[deadIndex].dead = true;

	// 점수 누적
	score += deadIndex;

	// 도둑말 이동
	for (int i = 1; i <= 16; i++)
	{
		CHESS horse = tmpChess[i];

		if (horse.dead == true) continue;

		while (1)
		{
			int nr, nc, dir;

			dir = horse.dir;
			nr = horse.r + dr[dir];
			nc = horse.c + dc[dir];

			if ((nr == pt.r && nc == pt.c)
				|| (nr < 0 || nc < 0 || nr > 3 || nc > 3))
			{
				//    -, ↑, ↖, ←, ↙, ↓, ↘, →, ↗
				// => -, ↖, ←, ↙, ↓, ↘, →, ↗, ↑
				int changeDir[9] = { 0, 2, 3, 4, 5, 6, 7, 8, 1 };
				int nextDir = changeDir[dir];

				horse.dir = nextDir;
				tmpChess[i].dir = nextDir;

				continue;
			}
			else
			{
				int tmp;
				int changeIndex = tmpMAP[nr][nc];

				// 좌표만 변경
				tmp = tmpChess[i].r;
				tmpChess[i].r = tmpChess[changeIndex].r;
				tmpChess[changeIndex].r = tmp;

				tmp = tmpChess[i].c;
				tmpChess[i].c = tmpChess[changeIndex].c;
				tmpChess[changeIndex].c = tmp;

				// MAP에서 index 변경
				tmp = tmpMAP[nr][nc];
				tmpMAP[nr][nc] = tmpMAP[horse.r][horse.c];
				tmpMAP[horse.r][horse.c] = tmp;

				break;
			}
		}
	}

	// 술래 1 ~ 3칸 이동
	int sr, sc;

	sr = pt.r;
	sc = pt.c;
	for (int i = 1; i <= 3; i++)
	{
		int nr, nc;

		nr = sr + dr[pt.dir] * i;
		nc = sc + dc[pt.dir] * i;

		if (nr < 0 || nc < 0 || nr > 3 || nc > 3)
		{
			if (maxAnswer < score) maxAnswer = score;

			break;
		}

		pt.r = nr;
		pt.c = nc;

		DFS(pt, tmpMAP, tmpChess, score);
	}
}

int main()
{
	// scanf("%d", &T);
	T = 1;
	for (int tc = 1; tc <= T; tc++)
	{
		input();

		maxAnswer = 0;

		tagger.r = tagger.c = 0;

		DFS(tagger, MAP, chess, 0);

		printf("%d\n", maxAnswer);
	}

	return 0;
}

File: Python/섹션08. A형 문제 풀이 (~2022 상반기)/29. 술래잡기 체스 (코드트리, 2020 상반기 오전 2번).py
import copy

MAX = 4 + 2
MAX_H = 16 + 5 # 도둑말의 수

MAP = [[0] * MAX for _ in range(MAX)]

class CHESS:
    def __init__(self, r: int, c: int, dir: int, dead: bool):
        self.r = r
        self.c = c
        self.dir = dir
        self.dead = False

chess = [CHESS(0, 0, 0, False) for _ in range(MAX_H)]
tagger = CHESS(0, 0, 0, False)

# -, ↑, ↖, ←, ↙, ↓, ↘, →, ↗
dr = [0, -1, -1, 0, 1, 1, 1, 0, -1]
dc = [0, 0, -1, -1, -1, 0, 1, 1, 1]

def input_data():
    for r in range(4):
        line = list(map(int, input().split()))
        for c in range(4):
            index = line[c * 2]
            dir = line[c * 2 + 1]
                        
            MAP[r][c] = index
            
            chess[index].r = r
            chess[index].c = c
            chess[index].dir = dir
            chess[index].dead = False

def print_status(tagger, map_, chess, score): # for debug
    print(f"score : {score}")
    print(f"tagger : r {tagger.r}, c {tagger.c}, dir {tagger.dir}")

    print("live chess : ", end='')
    for i in range(1, 17):
        if not chess[i].dead:
            print(f"{i}, ", end='')
    print("")

    for r in range(4):
        for c in range(4):
            if r == tagger.r and c == tagger.c:
                print(f"({-1}, {tagger.dir})", end=' ')
            else:
                piece_num = map_[r][c]
                print(f"({piece_num}, {chess[piece_num].dir})", end=' ')
        print("")
    print("")

def DFS(prev_tagger, prev_map, prev_chess, score):
    global max_answer

    pt = copy.deepcopy(prev_tagger)
    tmp_map = copy.deepcopy(prev_map)
    tmp_chess = copy.deepcopy(prev_chess)

    dead_index = tmp_map[pt.r][pt.c]
    
    # 이미 잡힌 도둑말인 경우
    if tmp_chess[dead_index].dead == True: return

    # 술래 방향 변경
    pt.dir = tmp_chess[dead_index].dir
    
    # 도둑말 사망 처리
    tmp_chess[dead_index].dead = True
    
    # 점수 누적
    score += dead_index

    # 도둑말 이동
    for i in range(1, 17):
        horse = copy.deepcopy(tmp_chess[i])
        
        if horse.dead == True: continue

        while True:
            dir = horse.dir
            nr = horse.r + dr[dir]
            nc = horse.c + dc[dir]

            if (nr == pt.r and nc == pt.c) or \
                (nr < 0 or nc < 0 or nr > 3 or nc > 3):
                #    -, ↑, ↖, ←, ↙, ↓, ↘, →, ↗
				# => -, ↖, ←, ↙, ↓, ↘, →, ↗, ↑
                change_dir = [0, 2, 3, 4, 5, 6, 7, 8, 1]
                next_dir = change_dir[dir]
                
                horse.dir = next_dir
                tmp_chess[i].dir = next_dir                
                
                continue
            else:
                change_index = tmp_map[nr][nc]
                
                # 좌표만 변경
                tmp_chess[i].r, tmp_chess[change_index].r = tmp_chess[change_index].r, tmp_chess[i].r
                tmp_chess[i].c, tmp_chess[change_index].c = tmp_chess[change_index].c, tmp_chess[i].c
                              
                # MAP에서 index 변경
                tmp_map[nr][nc], tmp_map[horse.r][horse.c] = tmp_map[horse.r][horse.c], tmp_map[nr][nc]
                
                break

    # 술래 1 ~ 3칸 이동
    sr, sc = pt.r, pt.c    
    for i in range(1, 4):
        nr = sr + dr[pt.dir] * i
        nc = sc + dc[pt.dir] * i

        if nr < 0 or nc < 0 or nr > 3 or nc > 3:
            if max_answer < score: max_answer = score

            break

        pt.r, pt.c = nr, nc
        
        DFS(pt, tmp_map, tmp_chess, score)
        
# T = int(input())
T = 1
for tc in range(T): 
    global max_answer
    
    input_data()
    
    max_answer = 0
    
    tagger.r, tagger.c = 0, 0
    
    DFS(tagger, MAP, chess, 0)
    
    print(max_answer)